src/ServiceControl/MessageFailures/Api/EditFailedMessages.cs

```diff
@@ -0,0 +1,151 @@
+                    return HttpStatusCode.BadRequest;
+                }
+
+                if (string.IsNullOrWhiteSpace(edit.MessageBody) || edit.MessageHeaders == null)
+                {
+                    logging.WarnFormat("There is no message body on the edit-retry for MessageID {0}.", failedMessageId);
+                    return HttpStatusCode.BadRequest;
+                }
+
+                // Encode the body in base64 so that the new body doesn't have to be escaped
+                var base64String = Convert.ToBase64String(Encoding.UTF8.GetBytes(edit.MessageBody));
+                await Bus.SendLocal(new EditAndSend
+                {
+                    FailedMessageId = failedMessageId,
+                    NewBody = base64String,
+                    NewHeaders = edit.MessageHeaders.ToDictionary(kvp => kvp.Key, kvp => kvp.Value)
+                }).ConfigureAwait(false);
+
+                return HttpStatusCode.Accepted;
+            };
+        }
+
+        public LockedHeaderModificationValidator LockedHeaderModificationValidator { get; set; }
+
+        EditConfigurationModel GetEditConfiguration()
+        {
+            return new EditConfigurationModel
+            {
+                Enabled = Settings.AllowMessageEditing,
+                LockedHeaders = new[]
+                {
+                    Headers.MessageId,
+                    Headers.SagaId,
+                    Headers.CorrelationId,
+                    Headers.ControlMessageHeader,
+                    Headers.OriginatingSagaId,
+                    Headers.RelatedTo,
+                    Headers.ConversationId,
+                    Headers.MessageIntent,
+                    "NServiceBus.ExceptionInfo.ExceptionType",
+                    "NServiceBus.ExceptionInfo.HelpLink",
+                    "NServiceBus.ExceptionInfo.Message",
+                    "NServiceBus.ExceptionInfo.Source",
+                    "NServiceBus.ExceptionInfo.StackTrace"
+                },
+                SensitiveHeaders = new[]
+                {
+                    Headers.RouteTo,
+                    Headers.DestinationSites,
+                    Headers.OriginatingSite,
+                    Headers.HttpTo,
+                    Headers.ReplyToAddress,
+                    Headers.NServiceBusVersion,
+                    Headers.ReturnMessageErrorCodeHeader,
+                    Headers.IsSagaTimeoutMessage,
+                    Headers.IsDeferredMessage,
+                    Headers.SagaType,
+                    Headers.OriginatingSagaType,
+                    Headers.DelayedRetries,
+                    Headers.DelayedRetriesTimestamp,
+                    Headers.ImmediateRetries,
+                    Headers.ProcessingStarted,
+                    Headers.ProcessingEnded,
+                    Headers.TimeSent,
+                    "Header"
```

Comment:
is `header` expected to be a real key?

Reply:
Unfortunately it is. https://github.com/Particular/NServiceBus/blob/develop/src/NServiceBus.Core/Headers.cs#L61

Reply:
Scary 😱